app/web/features/auth/notifications/PushNotificationPermission.tsx

```diff
@@ -0,0 +1,161 @@
+  const classes = useStyles();
+
+  const [permission, setPermission] = useState(Notification.permission);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const hasPermission = permission === "granted";
+
+  const onPermissionGranted = async () => {
+    // If permission is granted, subscribe the user to push notifications
+    try {
+      // Check if service workers and push notifications are supported
+      if ("serviceWorker" in navigator && "PushManager" in window) {
+        const registration = await navigator.serviceWorker.register(
+          "/service-worker.js",
+          { scope: "/" }
+        );
+
+        const existingPushSubscription =
+          await registration.pushManager.getSubscription();
+        const p256dhKey = existingPushSubscription?.getKey("p256dh");
+        const { vapidPublicKey } = await getVapidPublicKey();
+
+        if (existingPushSubscription && p256dhKey) {
+          const publicKey = arrayBufferToBase64(p256dhKey);
+
+          /**
+           * The purpose of this check is to ensure that the push subscription is correctly authenticated with the server’s VAPID key.
+           * If the client’s p256dh key no longer matches the server’s vapidPublicKey, then the subscription is unsubscribed and needs
+           * to be re-registered to ensure the security and validity of the Web Push connection.
+           */
+          if (publicKey !== vapidPublicKey) {
+            await existingPushSubscription.unsubscribe();
+          } else {
+            return;
+          }
+        }
+
+        // Subscribe to push notifications using the PushManager API
+        const subscription: PushSubscription =
+          await registration.pushManager.subscribe({
+            userVisibleOnly: true,
+            applicationServerKey: vapidPublicKey,
+          });
+
+        await registerPushNotificationSubscription(subscription);
+      } else {
+        setErrorMessage(
+          t("notification_settings.push_notifications.error_unsupported")
+        );
+      }
+    } catch (error) {
+      setErrorMessage(
+        t("notification_settings.push_notifications.error_generic")
+      );
+    }
+  };
+
+  const turnPushNotificationsOn = async () => {
+    if (permission === "default") {
+      const result = await Notification.requestPermission();
+      setPermission(result);
+
+      if (result === "granted") {
+        await onPermissionGranted();
+      }
+    } else {
+      setErrorMessage(
+        t("notification_settings.push_notifications.error_blocked_push")
+      );
+    }
+  };
+
+  const turnPushNotificationsOff = async () => {
+    const registration = await navigator.serviceWorker.register(
+      "/service-worker.js",
+      { scope: "/" }
+    );
+    const existingPushSubscription =
+      await registration.pushManager.getSubscription();
+
+    if (existingPushSubscription) {
+      await existingPushSubscription.unsubscribe();
+    }
+    setPermission("default");
```

Comment:
Ohhh. I see. But now this is not actually what the `Notification.permission` setting would be?

If you re-sub you don't need to request again. But I guess `Notification.requestPermission` doesn't ask for perms again if it already has it, so that itself doesn't matter.

Reply:
Actually I think I am setting it to "default" but really it would be "denied".

But yes it won't ask again, I'll update the `setPermission` value.